### DISCUSSION TITLE
Using flask-socketio and handling multiple requests at the same time

### DISCUSSION THREAD

**Discussion by adhrames** [miguelgrinberg/Flask-SocketIO#2090] [Q&A]
import eventlet
eventlet.monkey_patch()

from server import app, socketio

if __name__ == '__main__':
    socketio.run(app)

This is the wsgi.py file, socket connections do connect but multiple requests aren't handled and the responses get mixed and the same response is sent to both requests.
gunicorn -k eventlet -w 2 -b 0.0.0.0:8765 wsgi:app, This is the run command I am using, and I am seeing other discussions where it's mentioned eventlet is not suitable anymore,
Can anyone give me proper direction on how to handle multiple requests while also having seperate socket connections to each of them?

**Comment by miguelgrinberg** [comment]
>  but multiple requests aren't handled and the responses get mixed 

I don't understand what you mean by this. There are no requests and responses in Socket.IO. I think this is likely a misunderstanding on your part on how Socket.IO works.

> gunicorn -k eventlet -w 2 -b 0.0.0.0:8765 wsgi:app

See the documentation for how to run Gunicorn with eventlet. The `-w 2` is not going to work, use `-w 1`.

> and I am seeing other discussions where it's mentioned eventlet is not suitable anymore,

This is true, but unrelated to the problems that you are having as far as I can see.

> Can anyone give me proper direction on how to handle multiple requests while also having seperate socket connections to each of them?

There is nothing special that needs to be done, and there is a working Socket.IO example in the `example` directory of this repository if you want to get started with working code.

**Reply by adhrames** [reply]
Maybe the terminology I am using is wrong, but creating 2 connections at the same time with the gunicorn server when it's running a flask-socketio app is causing issues. Basically my app works this way, 
In the front-end JS, it gets some input from the user accessing the webpage, ip_addr:9779, emits that to the server using socket.io emit function, On the backend the response is generated only after the js emits the data from the user, and that's sent back to the front-end JS.

> See the documentation for how to run Gunicorn with eventlet. The -w 2 is not going to work, use -w 1.

I did read the documentation and tried the -w 1 way, The issue is, when I am opening up different tabs of this ip_addr:9779, And therefore emitting 2 different messages by giving 2 different inputs, The responses from the server are getting mixed, Is this something to do with the sessions?

**Reply by miguelgrinberg** [reply]
This is likely a bug in your own application. Without seeing the code that emits these responses it is hard for me to guess what it is you are doing wrong.

**Reply by adhrames** [reply]
```python
from flask_socketio import SocketIO, emit
app=Flask(__name__)
CORS(app)
socketio = SocketIO(app, cors_allowed_origins="*", async_mode='eventlet', logger=True, engineio_logger=True, ping_timeout=30, ping_interval=25)

# From here multiple threads are created to reduce runtime and after each thread finish is sent as an update to the front-end client
@socketio.on('search_request')
def handle_search_request(data):
     
    keyword = data.get("key")
    images = data.get("check")
    search_type = data.get("chosen")

    if not data:
        socketio.emit('search_response', {'error': "Keyword not there!"})
        return
    
    global search_count
    search_count = 0
    acc_search_results_dict = {}
    lock = threading.Lock()

    commands = generate_search_commands(images, keyword, search_type)
    arguments = [
        (cmd["cmd"], cmd["key_string"], acc_search_results_dict, lock)
        for cmd in commands
    ]
    threads = []
    for arg in arguments:
        thread = threading.Thread(target=run_search_cmds, args=arg)
        threads.append(thread)
        thread.start()
    
    for thread in threads:
        thread.join()
        
    print(f"No. of results for the search parameters ({keyword}, [{images}], {search_type}) : {len(acc_search_results_dict)}")

    socketio.emit('search_response', acc_search_results_dict)

#The socket_search_update() function is run after each thread is complete
def socket_search_update(lock, key_string):
    global search_count
    with lock:
        timestamp = datetime.now().isoformat()
        print(f"{timestamp} - Emitting update: {search_count}")
        socketio.emit('search_update', {'search_count': search_count, 'update_string': key_string})
        socketio.sleep(0)
```

**Reply by miguelgrinberg** [reply]
This:

```python
    socketio.emit('search_response', acc_search_results_dict)
```

Is a broadcast. The message is sent to all connected clients. If you want to send only to the sender of the originating event you have to use the context version of emit:

```python
    emit('search_response', acc_search_results_dict)
```

Of if you prefer to continue using the generic emit, then add the `to` argument:

```python
    socketio.emit('search_response', acc_search_results_dict, to=request.sid)
```

Also a side note, your search threads cannot do any heavy CPU work, or else your eventlet application is going to block.

**Reply by adhrames** [reply]
Okay, So i tried this modification and also sent the request as an argument to the function where emit is again used.
The message is emitted, but on the client side, it's not being received.

My client-side code:
```js
ControllerViewModel.prototype._socketConnections = async function () {
  
    var socket = io('http://100.111.148.112:8765', {
      transports: ['websocket']  // Only allow WebSocket transport
  });

    socket.on('connect', async () => {
        console.log('Connected to Flask Socket.IO server');   
    });

    socket.on('search_update', async (statusUpdate) => {
      console.log(statusUpdate);
      console.log('Search count updated:', statusUpdate.search_count, 'at', new Date().toISOString());
      totalCount = this.finalProductCount();
      completionPercentage = Math.round((statusUpdate.search_count/totalCount)*100);

      const lastHyphenIndex = statusUpdate.update_string.lastIndexOf('_');
      const productName = statusUpdate.update_string.substring(0, lastHyphenIndex);  // Text before the last hyphen
      const searchType = statusUpdate.update_string.substring(lastHyphenIndex + 1); // Text after the last hyphen
      let searchprogressText = "Shiphome " + productName + " searched for the " + searchType;
      this.searchProgressString(searchprogressText);

      console.log(completionPercentage);
      this.searchProgress(completionPercentage);
      // Update the Knockout observable
      //
    });

    socket.on('search_response', async (dataFromService) => {
      this.searchProgress(100);
      this.searchResults(dataFromService);
      this.afterSocketResponse();
      console.log('Search completed!');
      if (dataFromService.error) {
          console.error(dataFromService.error);
      } else {
          console.log('Search completed!', dataFromService);
      }
    });

   }
```
   This function includes all the socket connection handling,
   
   On a different function which activated after the user clicks on a submit button on the ui,
```js
   var socket2 = io('http://100.111.148.112:8765');
    socket2.emit('search_request', jsonServiceRequest);
 ```
And regarding the CPU heavy work, what do you recommend? I am not very strict on the worker-class, I just want socket.io to work.

**Reply by miguelgrinberg** [reply]
I'm sorry, but I don't understand what you are showing me. You shouldn't have two different sockets, each client has to have one socket connection. I really think you should look at working examples of Socket.IO apps instead of trying to improvise a solution.

> And regarding the CPU heavy work, what do you recommend? I am not very strict on the worker-class, I just want socket.io to work.

I can't recommend anything. Eventlet does not work with CPU heavy tasks.

**Reply by adhrames** [reply]
> I can't recommend anything. Eventlet does not work with CPU heavy tasks.

Yeah, I get that. But, I am not fixed on eventlet. What should I be using instead?

**Reply by miguelgrinberg** [reply]
Eventlet is optional. Just don't use it. If you uninstall it from your virtualenv, then Flask-SocketIO is going to use standard threads for concurrency.
